java/core/src/test/java/sleeper/core/partition/PartitionSubtreeFactoryTest.java

```diff
@@ -0,0 +1,207 @@
+                .splitToNewChildren(L1_RIGHT, L2_LEFT_OF_L1R, L2_RIGHT_OF_L1R, 123456789L)
+                .splitToNewChildren(L2_LEFT_OF_L1L, L3_LEFT_OF_L2LL, L3_RIGHT_OF_L2LL, -2000000L)
+                .splitToNewChildren(L2_RIGHT_OF_L1L, L3_LEFT_OF_L2LR, L3_RIGHT_OF_L2LR, -500000L)
+                .splitToNewChildren(L2_LEFT_OF_L1R, L3_LEFT_OF_L2RL, L3_RIGHT_OF_L2RL, 12345678L)
+                .splitToNewChildren(L2_RIGHT_OF_L1R, L3_LEFT_OF_L2RR, L3_RIGHT_OF_L2RR, 234567890L)
+                .buildTree();
+
+        // When 1
+        PartitionTree largeSubtree = PartitionSubtreeFactory.createSubtree(level3TreeOriginal, largePartitionCount);
+
+        // Then 1
+        assertThat(largeSubtree.getLeafPartitions().size()).isEqualTo(largePartitionCount);
+        assertThat(largeSubtree).isEqualTo(new PartitionsBuilder(schema)
+                .rootFirst(ROOT)
+                .splitToNewChildren(ROOT, L1_LEFT, L1_RIGHT, 0L)
+                .splitToNewChildren(L1_LEFT, L2_LEFT_OF_L1L, L2_RIGHT_OF_L1L, -1000000L)
+                .splitToNewChildren(L1_RIGHT, L2_LEFT_OF_L1R, L2_RIGHT_OF_L1R, 123456789L)
+                .splitToNewChildren(L2_LEFT_OF_L1L, L3_LEFT_OF_L2LL, L3_RIGHT_OF_L2LL, -2000000L)
+                .splitToNewChildren(L2_RIGHT_OF_L1L, L3_LEFT_OF_L2LR, L3_RIGHT_OF_L2LR, -500000L)
+                .buildTree());
+
+        // When 2
+        PartitionTree midSubtree = PartitionSubtreeFactory.createSubtree(level3TreeOriginal, midPartitionCount);
+
+        // Then 2
+        assertThat(midSubtree.getLeafPartitions().size()).isEqualTo(midPartitionCount);
+        assertThat(midSubtree).isEqualTo(new PartitionsBuilder(schema)
+                .rootFirst(ROOT)
+                .splitToNewChildren(ROOT, L1_LEFT, L1_RIGHT, 0L)
+                .splitToNewChildren(L1_LEFT, L2_LEFT_OF_L1L, L2_RIGHT_OF_L1L, -1000000L)
+                .splitToNewChildren(L1_RIGHT, L2_LEFT_OF_L1R, L2_RIGHT_OF_L1R, 123456789L)
+                .buildTree());
+
+        // When 3
+        PartitionTree smallSubtree = PartitionSubtreeFactory.createSubtree(level3TreeOriginal, smallPartitionCount);
+
+        // Then 3
+        assertThat(smallSubtree.getLeafPartitions().size()).isEqualTo(smallPartitionCount);
+        assertThat(smallSubtree).isEqualTo(new PartitionsBuilder(schema)
+                .rootFirst(ROOT)
+                .splitToNewChildren(ROOT, L1_LEFT, L1_RIGHT, 0L)
+                .buildTree());
+
+    }
+
+    @Test
+    void shouldThrowExceptionIfRequestLeafCountGreaterThanOriginalTree() {
+        assertThatThrownBy(() -> PartitionSubtreeFactory.createSubtree(new PartitionsBuilder(schema)
+                .rootFirst(ROOT)
+                .splitToNewChildren(ROOT, L1_LEFT, L1_RIGHT, 0L)
+                .buildTree(),
+                7))
+                .isInstanceOf(PartitionTreeException.class)
+                .hasMessageContaining("Requested size of 7 is greater than");
```

Comment:
Maybe it would be best to return the original tree in this case. I can see a case where we have a big EMR cluster and a lot of Sleeper tables, and we want to target the number of EMR nodes, but some Sleeper tables have fewer partitions than that.